src/main/java/build/buf/protovalidate/CustomOverload.java

```diff
@@ -611,40 +656,583 @@ private static boolean validateURI(String val, boolean checkAbsolute) {
-   * @param ver The IP version to validate against (0 for any version, 4 for IPv4, 6 for IPv6).
-   * @param strict If strict is true and host bits are set in the supplied address, then false is
-   *     returned.
-   * @return {@code true} if the input string is a valid IP prefix of the specified version, {@code
-   *     false} otherwise.
+   * <p>An address prefix divides an IP address into a network portion, and a host portion. The
+   * prefix length specifies how many bits the network portion has. For example, the IPv6 prefix
+   * "2001:db8:abcd:0012::0/64" designates the left-most 64 bits as the network prefix. The range of
+   * the network is 2**64 addresses, from 2001:db8:abcd:0012::0 to
+   * 2001:db8:abcd:0012:ffff:ffff:ffff:ffff.
+   *
+   * <p>An address prefix may include a specific host address, for example
+   * "2001:db8:abcd:0012::1f/64". With strict = true, this is not permitted. The host portion must
+   * be all zeros, as in "2001:db8:abcd:0012::0/64".
+   *
+   * <p>The same principle applies to IPv4 addresses. "192.168.1.0/24" designates the first 24 bits
+   * of the 32-bit IPv4 as the network prefix.
+   */
+  static boolean isIPPrefix(String str, long version, boolean strict) {
+    if (version == 6L) {
+      Ipv6 ip = new Ipv6(str);
+      return ip.addressPrefix() && (!strict || ip.isPrefixOnly());
+    } else if (version == 4L) {
+      Ipv4 ip = new Ipv4(str);
+      return ip.addressPrefix() && (!strict || ip.isPrefixOnly());
+    } else if (version == 0L) {
+      return isIPPrefix(str, 6, strict) || isIPPrefix(str, 4, strict);
+    }
+    return false;
+  }
+}
+
+final class Ipv4 {
+  private String str;
+  private int index;
+  private List<Short> octets;
+  private long prefixLen;
+
+  Ipv4(String str) {
+    this.str = str;
+    this.octets = new ArrayList<Short>();
+  }
+
+  /**
+   * Returns the 32-bit value of an address parsed through address() or addressPrefix().
+   *
+   * <p>Note Java does not support unsigned numeric types, so to handle unsigned 32-bit values, we
+   * need to use a 64-bit long type instead of the 32-bit (signed) Integer type.
+   *
+   * <p>Returns 0 if no address was parsed successfully.
+   */
+  int getBits() {
+    if (this.octets.size() != 4) {
+      return -1;
+    }
+    return (this.octets.get(0) << 24)
+        | (this.octets.get(1) << 16)
+        | (this.octets.get(2) << 8)
+        | this.octets.get(3);
+  }
+
+  /**
+   * Returns true if all bits to the right of the prefix-length are all zeros.
+   *
+   * <p>Behavior is undefined if addressPrefix() has not been called before, or has returned false.
    */
-  private static boolean validateIPPrefix(String prefix, long ver, boolean strict) {
-    IPAddressString str;
-    IPAddress addr;
+  boolean isPrefixOnly() {
+    int bits = this.getBits();
+
+    int mask = 0;
+    if (this.prefixLen == 32) {
+      mask = 0xffffffff;
+    } else {
+      mask = ~(0xffffffff >>> this.prefixLen) >>> 0;
```

Comment:
What is the `>>> 0` for here? I think it can be removed.

If we're pulling in Guava code I'd feel much more comfortable using their utilities to parse IP addresses instead of rolling our owns if they pass our tests.